agent/engine/docker_task_engine.go

```diff
@@ -1238,12 +1240,139 @@ func (engine *DockerTaskEngine) GetDaemonManagers() map[string]dm.DaemonManager
 func (engine *DockerTaskEngine) pullContainerManifest(
 	task *apitask.Task, container *apicontainer.Container,
 ) dockerapi.DockerContainerMetadata {
-	// Currently a no-op
-	logger.Debug("Manifest pull is currently a no-op", logger.Fields{
-		field.TaskID:    task.GetID(),
-		field.Container: container.Name,
-		field.Image:     container.Image,
+	switch container.Type {
+	case apicontainer.ContainerCNIPause,
+		apicontainer.ContainerNamespacePause,
+		apicontainer.ContainerServiceConnectRelay,
+		apicontainer.ContainerManagedDaemon:
+		// pause images and AppNet relay image are managed at startup
+		return dockerapi.DockerContainerMetadata{}
+	}
+	// AppNet Agent container image is also managed at start up
+	// (it uses the same image as AppNet Relay container)
+	if task.IsServiceConnectEnabled() && container == task.GetServiceConnectContainer() {
+		return dockerapi.DockerContainerMetadata{}
+	}
+
+	var imageManifestDigest digest.Digest
+	digestFromPayload := referenceutil.GetDigestFromImageRef(container.Image)
+	if digestFromPayload != "" {
+		// Digest available in task payload
+		imageManifestDigest = digestFromPayload
+		logger.Info("Found image manifest digest in task payload", logger.Fields{
+			field.TaskARN:       task.Arn,
+			field.ContainerName: container.Name,
+			field.Image:         container.Image,
+		})
+	} else if !engine.imagePullRequired(engine.cfg.ImagePullBehavior, container, task.GetID()) {
+		// Digest should be resolved by inspecting local image
```

Comment:
could you add a little more explanation to this comment on why this case exists? IIUC is it only for locally cached images and if agent is configured to prefer cached?

Reply:
Yeah I can do that. The idea is that resolved digest should match the digest of the image which would be used for the container. If image pull is not required then Agent attempts to use a locally available image for the container during transition to CREATED state.

Reply:
Updated.